packages/go_router/lib/src/misc/stacked_navigation_scaffold.dart

```diff
@@ -0,0 +1,207 @@
+    required this.stackItems,
+    this.scaffoldBuilder,
+    this.transitionBuilder,
+    this.transitionDuration = defaultTransitionDuration,
+    super.key,
+  });
+
+  /// The default transition duration
+  static const Duration defaultTransitionDuration = Duration(milliseconds: 400);
+
+  /// The navigator for the currently active tab
+  final Navigator currentNavigator;
+
+  /// The current router state
+  final GoRouterState currentRouterState;
+
+  /// The tabs
+  final List<StackedNavigationItem> stackItems;
+
+  /// The scaffold builder
+  final StackedNavigationScaffoldBuilder? scaffoldBuilder;
+
+  /// An optional transition builder for stack transitions
+  final StackedNavigationTransitionBuilder? transitionBuilder;
+
+  /// The duration for stack transitions
+  final Duration transitionDuration;
+
+  @override
+  State<StatefulWidget> createState() => _StackedNavigationScaffoldState();
+}
+
+class _StackedNavigationScaffoldState extends State<StackedNavigationScaffold>
+    with SingleTickerProviderStateMixin {
+  int _currentIndex = 0;
+  late final AnimationController? _animationController;
+  late final List<StackedNavigationItemState> _items;
+
+  int _findCurrentIndex() {
+    final int index = _items.indexWhere((StackedNavigationItemState i) =>
+        i.item.navigatorKey == widget.currentNavigator.key);
+    return index < 0 ? 0 : index;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _items = widget.stackItems
+        .map((StackedNavigationItem i) => StackedNavigationItemState(i))
+        .toList();
+
+    if (widget.transitionBuilder != null) {
+      _animationController =
+          AnimationController(vsync: this, duration: widget.transitionDuration);
+      _animationController?.forward();
+    } else {
+      _animationController = null;
+    }
+  }
+
+  @override
+  void didUpdateWidget(covariant StackedNavigationScaffold oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _updateForCurrentTab();
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _updateForCurrentTab();
+  }
+
+  void _updateForCurrentTab() {
+    final int previousIndex = _currentIndex;
+    _currentIndex = _findCurrentIndex();
+
+    final StackedNavigationItemState itemState = _items[_currentIndex];
+    itemState.navigator = widget.currentNavigator;
+    itemState.currentRouterState = widget.currentRouterState;
+
+    if (previousIndex != _currentIndex) {
+      _animationController?.forward(from: 0.0);
+    }
+  }
+
+  @override
+  void dispose() {
+    _animationController?.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final StackedNavigationScaffoldBuilder? scaffoldBuilder =
+        widget.scaffoldBuilder;
+    if (scaffoldBuilder != null) {
+      return scaffoldBuilder(
+          context, _currentIndex, _items, _buildIndexStack(context));
+    } else {
+      return _buildIndexStack(context);
```

Comment:
One thing I found slightly confusing is that this widget is called `StackedNavigationScaffold`, but if no `scaffoldBuilder` is given we only return an `IndexedStack`.

Can't think of a better name though.

Reply:
True. I was actually considering the name `StackedNavigationShell` instead.

Reply:
Yeah, that seems more suitable.